### DISCUSSION TITLE
What is the preferred way to capture data from notifications or UART RX?

### DISCUSSION THREAD

**Discussion by ch701x** [ThomasGerstenberg/blatann#122] [Q&A] [answered]
Taking the `central_uart_service.py` example, the response from the peripheral is output through the function `on_data_rx`. Let's say I want to capture this data in a queue and perform operations on it in the main thread. I found I can use `time.sleep` to allow the event thread to put received data in the queue, but it seems like there should be a more event based solution to this. Rather than sleeping for an arbitrary amount of time.

So my question is, what is the preferred or expected way to capture data from event threads and do operations on said data?

Thank you.

**Comment by ThomasGerstenberg** [comment, marked as answer]
You could use something like a [queue](https://docs.python.org/3/library/queue.html) to put the bytes received in the callback, then wait on that queue in the main thread. That gives you thread synchronization so when a notification with serial data is received and put in the queue, it'll wake up the main thread (or wherever called `queue.get()`.

Depending on the application, you can queue the whole notification packet as a single entity, queue individual bytes, or if it's text you could parse the data and queue individual lines. The individual byte solution is less performant and there are likely better options there (socket, os.pipe maybe, etc.).
